pep-0484.txt

```diff
@@ -861,6 +861,97 @@ allow all operations on it, and a value of type ``Any`` can be assigned
 to a variable (or used as a return value) of a more constrained type.
 
 
+The type of class objects
+-------------------------
+
+Sometimes you want to talk about class objects, in particular class
+objects that inherit from a given class.  This can be spelled as
+``Type[C]`` where ``C`` is a class.  To clarify: while ``C`` (when
+used as an annotation) refers to instances of class ``C``, ``Type[C]``
+refers to *subclasses* of ``C``.  (This is a similar distinction as
+between ``object`` and ``type``.)
+
+For example, suppose we have the following classes::
+
+  class User: ...  # Abstract base for User classes
+  class BasicUser(User): ...
+  class ProUser(User): ...
+  class TeamUser(User): ...
+
+And suppose we have a function that creates an instance of one of
+these classes if you pass it a class object::
+
+  def new_user(user_class):
+      user = user_class()
+      # (Here we could write the user object to a database)
+      return user
+
+Without ``Type[]`` the best we could do to annotate ``new_user()``
+would be::
+
+  def new_user(user_class: type) -> User:
+      ...
+
+However using ``Type[]`` and a type variable with an upper bound we
+can do much better::
+
+  U = TypeVar('U', bound=User)
+  def new_user(user_class: Type[U]) -> U:
+      ...
+
+Now when we call ``new_user()`` with a specific subclass of ``User`` a
+type checker will infer the correct type of the result::
+
+  joe = new_user(BasicUser)  # Inferred type is BasicUser
+
+The value corresponding to ``Type[C]`` must be an actual class object
+that's a subtype of ``C``, not a special form.  IOW, in the above
+example calling e.g. ``new_user(Union[BasicUser, ProUser])`` is
+rejected by the type checker (in addition to failing at runtime
+because you can't instantiate a union).
+
+Note that it *is* legal to use a union as the parameter for
+``Type[]``, as in::
+
+  def new_non_team_user(user_class: Type[Union[BasicUser, ProUser]]):
+      ...
+
+However the actual argument passed in at runtime must still be a
+concrete class object, e.g.::
+
+  new_non_team_user(ProUser)  # OK
+  new_non_team_user(TeamUser)  # Disallowed
```

Comment:
This is probably a really minor comment, by I would replace "Disallowed" by "Fails type check" or similar, since at first I was confused that it is disallowed because ``TeamUser`` is not a concrete class object. But maybe it is just me.